src/sagas.js

```diff
@@ -164,39 +176,66 @@ function* fetchIcon ({ iconId }) {
 }
 
 function* refreshData (action) {
-  console.log('refreshData(): ', action.payload)
+  logDebug('refreshData(): ', action.payload)
   if (!action.payload.quiet) {
-    logDebug('refreshData(): not quiet')
     yield put(loadInProgress({ value: true }))
   }
 
-  // do refresh sequentially
-  yield fetchVmsByCount(getVmsByCount({ count: action.payload.page * AppConfiguration.pageLimit, shallowFetch: !!action.payload.shallowFetch }))
-  yield fetchPoolsByCount(getPoolsByCount({ count: action.payload.page * AppConfiguration.pageLimit }))
+  // refresh VMs and remove any that haven't been refreshed
+  const vmsFetchedById = []
+  for (let vmId of Selectors.getVmsFetchedById().toArray()) {
+    const status = yield fetchSingleVm(getSingleVm({ vmId }))
+    if (status) vmsFetchedById.push(vmId)
+  }
+
+  const fetchedVmIds = yield fetchVmsByCount(getVmsByCount({
+    count: action.payload.page * AppConfiguration.pageLimit,
+    shallowFetch: !!action.payload.shallowFetch,
+  }))
+
+  yield put(removeMissingVms({ vmIdsToPreserve: [ ...fetchedVmIds, ...vmsFetchedById ] }))
+  yield put(removeVmsFetchedById({ vmIds: fetchedVmIds }))
 
+  // refresh Pools and remove any that haven't been refreshed
+  const poolsFetchedById = []
+  for (let poolId of Selectors.getPoolsFetchedById().toArray()) {
+    const status = yield fetchSinglePool(getSinglePool({ poolId }))
+    if (status) poolsFetchedById.push(poolId)
+  }
+
+  const fetchedPoolIds = yield fetchPoolsByCount(getPoolsByCount({
+    count: action.payload.page * AppConfiguration.pageLimit,
+  }))
+
+  yield put(removeMissingPools({ poolIdsToPreserve: [ ...fetchedPoolIds, ...poolsFetchedById ] }))
+  yield put(removePoolsFetchedById({ poolIds: fetchedPoolIds }))
+
+  // update counts
+  yield put(updateVmsPoolsCount())
   yield put(loadInProgress({ value: false }))
```

Comment:
the `loadInProgress` refactoring was just merged, please review.

Reply:
Rebase will address this